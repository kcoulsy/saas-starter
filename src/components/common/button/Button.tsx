import Spinner from '../spinner/Spinner';
import { cn } from '@src/utils/cn';

export interface ButtonProps {
  id?: string;
  label: string;
  type: 'submit' | 'button' | 'reset';
  classes?: string;
  onClick?: () => void;
  disabled?: boolean;
  isLoading?: boolean;
}

const Button = ({ id, type = 'button', label, onClick, classes, disabled, isLoading }: ButtonProps) => {
  return (
    <button
      id={id}
      // eslint-disable-next-line react/button-has-type
      type={type}
      onClick={!disabled ? onClick : undefined}
      aria-label={label}
      className={cn(
        `text-sm font-semibold leading-none text-white focus:outline-none flex justify-center items-center h-10 w-full`,
        {
          'bg-indigo-700 border rounded hover:bg-indigo-600 ': !disabled,
          'focus:ring-2 focus:ring-offset-2 focus:ring-indigo-700': !disabled,
          'bg-gray-300 border rounded cursor-not-allowed': disabled,
        },
        classes,
      )}
    >
      {isLoading ? <Spinner classes="h-5 w-5" /> : label}
    </button>
  );
};

export default Button;
