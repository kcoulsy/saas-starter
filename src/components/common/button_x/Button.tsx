import clsx from 'clsx';

export interface ButtonProps {
  id?: string;
  label: string;
  type: 'submit' | 'button' | 'reset';
  classes?: string;
  onClick?: () => void;
  disabled?: boolean;
}

const Button = ({ id, type = 'button', label, onClick, classes, disabled }: ButtonProps) => {
  return (
    <button
      id={id}
      // eslint-disable-next-line react/button-has-type
      type={type}
      onClick={!disabled ? onClick : undefined}
      aria-label={label}
      className={clsx(`text-sm font-semibold leading-none text-white focus:outline-none py-4 w-full ${classes}`, {
        'bg-indigo-700 border rounded hover:bg-indigo-600 ': !disabled,
        'focus:ring-2 focus:ring-offset-2 focus:ring-indigo-700': !disabled,
        'bg-gray-300 border rounded cursor-not-allowed': disabled,
      })}
    >
      {label}
    </button>
  );
};

export default Button;
