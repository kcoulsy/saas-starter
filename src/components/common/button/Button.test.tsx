import { composeStories } from '@storybook/react';
import { render, screen } from '@testing-library/react';
import * as stories from './Button.stories';
import userEvent from '@testing-library/user-event';

const { Default, Disabled, Loading } = composeStories(stories);

describe('Button', () => {
  it('should render and match snapshot', () => {
    const { asFragment } = render(<Default />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render disabled', () => {
    const { asFragment } = render(<Disabled />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render with custom class', () => {
    render(<Default classes="custom-class" />);

    const button = screen.getByRole('button');

    expect(button).toHaveClass('custom-class');
  });

  it('should render with custom text', () => {
    render(<Default label="Custom text" />);

    const button = screen.getByRole('button');

    expect(button).toHaveTextContent('Custom text');
  });

  it('should call onClick when clicked', async () => {
    const onClick = vi.fn().mockImplementation(() => {});

    render(<Default onClick={onClick} />);

    const button = screen.getByRole('button');

    await userEvent.click(button);

    expect(onClick).toHaveBeenCalled();
  });

  it('should not call onClick if the button is disabled', async () => {
    const onClick = vi.fn().mockImplementation(() => {});

    render(<Disabled onClick={onClick} />);

    const button = screen.getByRole('button');

    await userEvent.click(button);

    expect(onClick).not.toHaveBeenCalled();
  });

  it('should render spinner when isLoading is true', () => {
    render(<Loading />);

    const button = screen.getByRole('button');
    const spinner = screen.getByRole('status');

    expect(button).toContainElement(spinner);
  });
});
