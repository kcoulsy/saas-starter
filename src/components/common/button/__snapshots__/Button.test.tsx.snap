// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button > should render and match snapshot 1`] = `
<DocumentFragment>
  <div
    class="w-40"
  >
    <button
      aria-label="Submit"
      class="text-sm font-semibold leading-none text-white focus:outline-none flex justify-center items-center h-10 w-full bg-indigo-700 border rounded hover:bg-indigo-600 focus:ring-2 focus:ring-offset-2 focus:ring-indigo-700"
      id="button-base"
      type="button"
    >
      Submit
    </button>
  </div>
</DocumentFragment>
`;

exports[`Button > should render disabled 1`] = `
<DocumentFragment>
  <div
    class="w-40"
  >
    <button
      aria-label="Submit"
      class="text-sm font-semibold leading-none text-white focus:outline-none flex justify-center items-center h-10 w-full bg-gray-300 border rounded cursor-not-allowed"
      id="button-disabled"
      type="button"
    >
      Submit
    </button>
  </div>
</DocumentFragment>
`;
